src/metrics/misc.ts

```diff
@@ -207,39 +245,125 @@ export class MetadataMetric implements Metric {
     } else if (result === '[Result "1/2-1/2"]') {
       this.ties++;
     } else {
-      throw new Error('Invalid result');
+      console.log('Invalid result');
     }
 
-    // extract openings from metadata
-    const opening = metadata?.find((item) => item.startsWith('[Opening "'))
-    ?.replace('[Opening "', '')
-    ?.replace('"]', '');
+// extract openings from metadata
+const opening = metadata?.find((item) => item.startsWith('[Opening "'))
+?.replace('[Opening "', '')
+?.replace('"]', '');
 
-    if (opening.toLowerCase() == "bongcloud") {
-      this.bongcloud++;
-    }
+if(opening) {
+  if (opening.toLowerCase() == "bongcloud") {
+    this.bongcloud++;
+  }
 
-    // add opening to openings object
-    if (this.openings[opening]) {
-      this.openings[opening].appearances++;
-      if (result === '[Result "1-0"]') {
+  // add opening to openings object
+  if (this.openings[opening]) {
+    this.openings[opening].appearances++;
+    switch(result) {
+      case '[Result "1-0"]':
         this.openings[opening].whiteWins++;
-      } else if (result === '[Result "0-1"]') {
+        break;
+      case '[Result "0-1"]':
         this.openings[opening].blackWins++;
-      } else if (result === '[Result "1/2-1/2"]') {
+        break;
+      case '[Result "1/2-1/2"]':
         this.openings[opening].whiteWins += 0.5;
         this.openings[opening].blackWins += 0.5;
         this.openings[opening].ties++;
+        break;
+    }
+    this.openings[opening].whiteToBlackWinRatio = this.openings[opening].whiteWins / this.openings[opening].blackWins;
+  } else {
+    let blackWins, whiteWins, ties, whiteToBlackWinRatio;
+
+    switch(result) {
+      case '[Result "0-1"]':
+        blackWins = 1;
+        whiteWins = 0;
+        ties = 0;
+        whiteToBlackWinRatio = 0;
+        break;
+      case '[Result "1-0"]':
+        blackWins = 0;
+        whiteWins = 1;
+        ties = 0;
+        whiteToBlackWinRatio = Infinity;
+        break;
+      case '[Result "1/2-1/2"]':
+        blackWins = 0.5;
+        whiteWins = 0.5;
+        ties = 1;
+        whiteToBlackWinRatio = 1;
+        break;
+      default:
+        blackWins = 0;
+        whiteWins = 0;
+        ties = 0;
+        whiteToBlackWinRatio = 0;
+    }
+
+    this.openings[opening] = {
+      appearances: 1,
+      blackWins: blackWins,
+      whiteWins: whiteWins,
+      ties: ties,
+      whiteToBlackWinRatio: whiteToBlackWinRatio,
+    };
+  }
+}
+
+
+    // identify game endings
+    let gameEnd = metadata.find(item => item.startsWith('[Termination'));
+    const lastMove = game[game.length - 1].move;
+
+    if (gameEnd === '[Termination "Normal"]') {
+      if (lastMove.originalString.includes('#')) {
+        this.gameEndings['checkmate'] = (this.gameEndings['checkmate'] || 0) + 1;
+      } else if (result === '[Result "1/2-1/2"]') {
+        this.gameEndings['draw'] = (this.gameEndings['draw'] || 0) + 1;
+        if (this.chess.isStalemate()) {
```

Comment:
I'm having trouble seeing where this.chess is instantiated with the board state. Recreating the board is also a little inefficient, but ik we need to understand the ending reason somehow, I'd need to look at the code again but maybe we can do it in a different place? I.e. when we get to the last move of a game history generator?